### DISCUSSION TITLE
Citation key formula with "etal" for 3 or more authors

### DISCUSSION THREAD

**Discussion by katlund** [retorquere/zotero-better-bibtex#2474] [Q&A] [answered]
I would like to autogenerate citation keys for my entire library based on the following rules:

- First 3 letters of first author's surname: Aaa
- First letter of the next two authors' surnames: B, C
- "etal" appended for 4 or more authors
- last two digits of publication year

Examples for different scenarios in 2023:
- Single author: Aaa23
- Two authors: AaaB23
- Three authors: AaaBC23
- Four or more authors: AaaBCetal23
- postfix for duplicates

The closest I've come is `auth(m=1, n=3) + auth(m=2, n=1) + auth(m=3, n=1) + shortyear` with the postfix option, which satisfies everything but the "etal" rule.  It seems all the built-in "etal" options start from 3 instead of 4 authors and don't allow for customizing the number of letters to take from author surnames.

Is there a way to encode this ruleset with existing features?

(By the way, the ruleset is not my own and must be complied with for historical reasons.)

**Comment by retorquere** [comment, marked as answer]
The following should work

```
authors(max=3).discard + auth(m=1, n=3) + auth(m=2, n=1) + auth(m=3, n=1) + shortyear | auth(m=1, n=3) + auth(m=2, n=1) + auth(m=3, n=1) + 'etal' + shortyear
```

**Reply by Zzaaii** [reply]
Hi, 
At the moment, I'm using it to export BibLaTex and then import to Obsidian.  However, when the citation key appears automatically on obsidian page as the title, it looks like a long sentence, if I won't change it manually. So, I wonder do you have simpler way to change this default? "auth.lower + first word of title + Year" is perfect for me. I changed the citation key formula to "auth.lower + shorttitle(1,1) + Year", as expected it didn't work...I checked the citation setting key page (https://retorque.re/zotero-better-bibtex/citing/), which is a bit complicated for me to make changes, due to my poor programming background. Or I would choose an earlier version to avoid this issue.

**Reply by retorquere** [reply]
Oh I see; you have capitalized `Year`. If you make that lowercase it will work.

**Reply by retorquere** [reply]
The new release will make the problem more visible, but in the version you have, if the field turns orange, you can hold the cursor still over the formula and it will show a text describing the problem with it.

**Reply by Zzaaii** [reply]
> The new release will make the problem more visible, but in the version you have, if the field turns orange, you can hold the cursor still over the formula and it will show a text describing the problem with it.

Hi,
thank you very much in checking up the details I have written. Now I have fixed the problem and even change it to the style I prefer a bit more. Here is the whole process that solved my issue after the checkup of formatting/typo:

1. I changed the "shorttitle" to "veryshorttitle" after reading Update README.md (https://github.com/retorquere/zotero-better-bibtex/commit/a39244e69cda22eb6258c88201367e9e3441b2e3). Didn't work.

2. Read Citation keys page again and found the critical information which I didn't notice before: "**existing keys are not automatically regenerated when you change the pattern. If you want your keys to update after a pattern change you will have to select your items, right-click, and select Refresh.**" (https://retorque.re/zotero-better-bibtex/citing/)

3. Action: Selected the items of references --> right click: At the bottom of the popped window, there's "Better BibTex" --> choose 'Refresh BibTex Key': Bingo!

4. Extra (For anyone who has interest): Get the format of "author_one word of shorttitle_year": I changed the default citation key formula to:
     `auth.lower + "_" + shorttitle(1,1) + "_" + year `

Thank you again for your help! :D 👍

**Reply by retorquere** [reply]
No worries, but I don't think it's correct to say step 1 didn't work. It worked, it just didn't do what you expected it to.

**Reply by Zzaaii** [reply]
> No worries, but I don't think it's correct to say step 1 didn't work. It worked, it just didn't do what you expected it to.

Ah, I see, you're right! And I didn't refresh to see the update. 🤦‍♀️

**Comment by katlund** [comment]
Now the etal rule is satisfied, but entries with 3 or fewer authors are
messed up, (I think) because `authors` prints the full surnames.  For
example: a 2023 entry with authors Aardvark, Bat, and Cardinal is assigned
AardvarkBatCardinalAarBC23, but it should just be AarBC23.

**Reply by retorquere** [reply]
See the updated formula. I forgot the `.discard`.

**Reply by katlund** [reply]
Perfect, thanks!
